### ISSUE TITLE
Extend RouteMatch to support matching based on the request source

### ISSUE THREAD

**Issue by georgi-d** [envoyproxy/envoy#4535]
We have a use case where for the same virtual host different the request needs to be routed differently based on the source of the request.

There are two different ways to distinguish the source:
1) Whether the the request comes from the same node where Envoy is running. Where for the request socket: `tcp.src_address == tcp.dst_address`. This check catches local connections on all interfaces.
2) The request's tcp.src_address belongs to a specific subnet ex. 192.168.0.0/24

One possible solution I see is to extend the RouteMatch structure with:

```
message RouteMatch {
  ...
  enun RequestSourceType {
    // Request is comming from withing the same network node as Envoy
    SAME_HOST = 0,
    // Request is comming from outside of the node
    EXTERNAL_HOST = 1,
  }
  oneof source {
    RequestSourceType source_type;
    repeated string source_subnet;
  } 
}
```

Thanks

**Comment by georgi-d** [comment]
@nickolaev

**Comment by mattklein123** [comment]
This sounds reasonable but please see how we are doing CIDR matching elsewhere in RBAC, filter chain matching, etc. We should use the same constructs.

**Comment by mattklein123** [comment]
Also, I'm not sure I agree we should support the "same host" semantics. IMO this can be configured by the operator with appropriate CIDR matches?

**Comment by georgi-d** [comment]
@mattklein123 

I assume you are talking about: https://www.envoyproxy.io/docs/envoy/v1.7.0/api-v2/config/rbac/v2alpha/rbac.proto

I am fine with using the same constructs as in RBAC for the matching.

The same host is crucial for my use case. I am not there is a way to achieve this without explicit support from envoy. It is not possible to pre-configure all the IPs that are on the host because interfaces and IPs can change at runtime and there is no stable way to monitor the changes. 

The simple check `src_ip == dest_ip` is guaranteed to work and correctly identify local connections.
There already is a functionality for this in Envoy: ` Network::Utility::isLocalConnection()`

**Comment by mattklein123** [comment]
OK fair enough re: same host. Sounds like it might be generally useful. Can you propose the final config changes so we can discuss if you are planning on doing the work?

**Comment by georgi-d** [comment]
I will think over it and propose final changes most likely tomorrow.

Thanks

**Comment by nickolaev** [comment]
After some discussions with @georgi-d we came up with the following proposal:

```
{
.....
  enum RequestSourceType {
    // Request does not originate from the node where Envoy is running
    EXTERNAL_HOST = 0;
    // Request originates from the node where Envoy is running
    SAME_HOST = 1;
  }

  message SourceMatch {
    oneof source_specifier {
      // Specifies the source IP match type. Can be same or external host
      RequestSourceType source_type = 10 [(validate.rules).enum.defined_only = true];
      // Specifies the optional external host subnet to match the sopurce IP.
      // Used when source_type is EXTERNAL_HOST
      envoy.api.v2.core.CidrRange source_subnet = 11;
    }
  }

  // Optionally specifies a source address match
  repeated SourceMatch source_matches = 12;
}
```

@mattklein123 what do you think?

**Comment by mattklein123** [comment]
This sounds reasonable to me. How are you going to concretely determine SAME_HOST? It should work regardless of address type and platform. cc @envoyproxy/maintainers

**Comment by nickolaev** [comment]
@htuch Wouldn't comparing `Network::Connection`'s `localAddress()` and `remoteAddress()` be sufficient enough in this case?

**Comment by htuch** [comment]
@nickolaev I'm curious if we can make this happen without any code changes today. The idea is you would have two distinct filter chains.

**Comment by georgi-d** [comment]
@htuch I have though about the separate listener option but it does not cover the same use case. We want to cover the case where the client makes a DNS request and gets the IP of the endpoint it needs to connect to. The IP it gets is the public address of the host and we need to catch this case. It is not possible to modify the DNS record to return localhost address and it would not be correct.

**Comment by georgi-d** [comment]
@htuch  There is one more reason a separate listener cannot be used. It is not possible to have one listener on localhost:80 and another listener on all other interfaces :80

**Comment by htuch** [comment]
@georgi-d sorry, I wasn't clear. The idea is that we have one listener but multiple filter chains. Each filter chain matches on distinct source IP criteria in its FilterChainMatch (e.g. local vs. non-local).

**Comment by georgi-d** [comment]
@htuch Having multiple filter chains based on the source address would also work. I like the idea so if it is the preffered solution by the rest of the team I am fine with switching to it.

**Comment by georgi-d** [comment]
@htuch After some more looking the source network functionality  is already there in `FilterChainMatch`. We only need to add the SameHost functionality.

**Comment by georgi-d** [comment]
I also see that the functionality has not been implemented

```
  // The criteria is satisfied if the source IP address of the downstream
  // connection is contained in at least one of the specified subnets. If the
  // parameter is not specified or the list is empty, the source IP address is
  // ignored.
  // [#not-implemented-hide:]
repeated core.CidrRange source_prefix_ranges = 6;
```

Since it has: `[#not-implemented-hide:]` and I do not see this parameter being used anywhere in the code

**Comment by mattklein123** [comment]
@georgi-d yeah would be great to get that implemented, it should be very straightforward. cc @PiotrSikora who might already be planning on implementing it. I'm not sure. (He has done the rest of the matching work in that area).

**Comment by htuch** [comment]
@georgi-d if that can work with you it seems this is ideal, as we then avoid implementing the functionality again in the route table. I agree with @mattklein123 that given the CIDR match support for destination VIPs, this should not be much work to do.

**Comment by georgi-d** [comment]
@htuch I will look into implementing the SAME_HOST functionality first since it is the feature I am blocked on. Will post a draft change when I have something working.

**Comment by nickolaev** [comment]
We're finalizing #4535, which should solve this issue. Shall we reopen it meanwhile?

**Comment by htuch** [comment]
Reopening, only partially fixed in #4682.

**Comment by mattklein123** [comment]
@htuch how is this issue only partially fixed? I'm cleaning up issues and got confused. We have a separate issue on IP matching? https://github.com/envoyproxy/envoy/issues/4457

**Comment by PiotrSikora** [comment]
My best guest is that #4682 implemented this at the filter chain level, but this issue is about implementing this also at the virtual host level (i.e. `RouteMatch`).

**Comment by htuch** [comment]
@PiotrSikora we agreed earlier in this issue that we didn't need to do it at virtual host level for @nickolaev 's needs, so if that's it, I think we can close.

**Comment by PiotrSikora** [comment]
@htuch ah, ok, I've missed that. We have at the filter chain level is complete, so I guess this can be closed.

**Comment by mattklein123** [comment]
OK closing.

**Comment by nickolaev** [comment]
I confirm the issue can be closed, as we already have a merged solution. #4457 will track the next developments.
Thank you all!
